akd/src/directory.rs

```diff
@@ -833,4 +833,188 @@ mod tests {
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn test_cleaned_user_state() -> Result<(), AkdError> {
+        let db = AsyncInMemoryDatabase::new();
+        let mut akd = Directory::<_>::new::<Blake3>(&db).await?;
+
+        // Publish twice
+        akd.publish::<Blake3>(
+            vec![
+                (AkdLabel("hello".to_string()), AkdValue("world".to_string())),
+                (
+                    AkdLabel("hello2".to_string()),
+                    AkdValue("world2".to_string()),
+                ),
+            ],
+            false,
+        )
+        .await?;
+
+        akd.publish::<Blake3>(
+            vec![
+                (AkdLabel("hello".to_string()), AkdValue("world".to_string())),
+                (
+                    AkdLabel("hello2".to_string()),
+                    AkdValue("world2".to_string()),
+                ),
+            ],
+            false,
+        )
+        .await?;
+
+        // Delete all user states from storage except most recent
+        db.clean_user_state(&[
+            AkdLabel("hello".to_string()),
+            AkdLabel("hello2".to_string()),
+        ])
+        .await;
+
+        // Verify lookup, key history and audit proofs
+        let current_azks = akd.retrieve_current_azks().await?;
+        let root_hash = akd.get_root_hash::<Blake3>(&current_azks).await?;
+
+        let history_proof = akd.key_history(&AkdLabel("hello".to_string())).await?;
```

Comment:
From what I can tell, the proof is constructed for the `ValueStates` that are retrieved. `get_user_data` would only return the latest `ValueState` when the previous ones have been deleted, and an update proof is only created for that `ValueState`. 

https://github.com/novifinancial/akd/blob/a409fc284865d4e69d51a218e262dbb94d25ef6e/akd/src/directory.rs#L241-L259

So wouldn't AKD avoid looking up `ValueStates`  that are already deleted and therefore not run into the "NotFound" errors?

Reply:
Hmm OK so that's no ideal. In that case it should be basing the updates off of the state of the node, not the user data retrieved. @Jasleen1 this is a change that will need to occur. There were updates at some epochs, but there just _happen_ to no longer be values associated to those updates. 

In the current implementation it's simply stating "there was no change at the given epoch".